cmd/restic/cmd_copy.go

```diff
@@ -0,0 +1,211 @@
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(gopts.ctx)
+	defer cancel()
+
+	srcRepo, err := OpenRepository(gopts)
+	if err != nil {
+		return err
+	}
+
+	dstRepo, err := OpenRepository(dstGopts)
+	if err != nil {
+		return err
+	}
+
+	srcLock, err := lockRepo(srcRepo)
+	defer unlockRepo(srcLock)
+	if err != nil {
+		return err
+	}
+
+	dstLock, err := lockRepo(dstRepo)
+	defer unlockRepo(dstLock)
+	if err != nil {
+		return err
+	}
+
+	Verbosef("Loading source index\n")
+	if err := srcRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	Verbosef("Loading destination index\n")
+	if err := dstRepo.LoadIndex(ctx); err != nil {
+		return err
+	}
+
+	for sn := range FindFilteredSnapshots(ctx, srcRepo, opts.Host, opts.Tags, opts.Paths, args) {
+		Verbosef("snapshot %s of %v at %s)\n", sn.ID().Str(), sn.Paths, sn.Time)
+
+		// TODO(): don't think this is sufficient, we prob need to iterate over
+		// the existing snapshots and do a full equivalance check.
+		if dstRepo.Index().Has(*sn.Tree, restic.TreeBlob) {
+			Printf("Snapshot %s already exists, skipped\n", sn.ID().Str())
+			continue
+		}
+		debug.Log("non-existence assurred")
+		Verbosef("  copy started, this may take a while...\n")
+
+		if err := copySnapshot(ctx, srcRepo, dstRepo, *sn.Tree); err != nil {
+			return err
+		}
+		debug.Log("tree copied")
+
+		if err = dstRepo.Flush(ctx); err != nil {
+			return err
+		}
+		debug.Log("flushed packs")
+
+		err = dstRepo.SaveIndex(ctx)
+		if err != nil {
+			debug.Log("error saving index: %v", err)
+			return err
+		}
+		debug.Log("saved index")
+
+		// save snapshot
+		sn.Parent = nil   // does Parent have relevance in the new repo?
+		sn.Original = nil // does Original have relevance in the new repo?
+		newid, err := dstRepo.SaveJSONUnpacked(ctx, restic.SnapshotFile, sn)
+		if err != nil {
+			return err
+		}
+		debug.Log("saved snapshot %v", newid.Str())
+	}
+	return nil
+}
+
+func copySnapshot(ctx context.Context, srcRepo, dstRepo restic.Repository, treeID restic.ID) error {
+	tree, err := srcRepo.LoadTree(ctx, treeID)
+	if err != nil {
+		return fmt.Errorf("LoadTree(%v) returned error %v", treeID.Str(), err)
+	}
+	// Do we already have this tree blob?
+	if !dstRepo.Index().Has(treeID, restic.TreeBlob) {
+		newtreeid, err := dstRepo.SaveTree(ctx, tree)
+		if err != nil {
+			return fmt.Errorf("SaveTree(%v) returned error %v", treeID.Str(), err)
+		}
+		// Assurance only.
+		if newtreeid != treeID {
+			return fmt.Errorf("SaveTree(%v) returned unexpected id %s", treeID.Str(), newtreeid.Str())
+		}
+	}
+
+	// TODO: keep only one (big) buffer around.
+	// TODO: parellize this stuff, likely only needed inside a tree.
+
+	for _, entry := range tree.Nodes {
+		// If it is a directory, recurse
+		if entry.Type == "dir" && entry.Subtree != nil {
+			if err := copySnapshot(ctx, srcRepo, dstRepo, *entry.Subtree); err != nil {
+				return err
+			}
+		}
+		if len(entry.Content) != 0 {
```

Comment:
This check is unnecessary. Removing it unindents the loop body by one tab.

Reply:
Done.

Not sure what noise you are referring to. All is well.